### PR TITLE
[reward, trainer] feat: support multi-output trajectories in async reward scoring

### PR DESCRIPTION
**Pull request by guillemgt**
### What does this PR do?

When using the TransferQueue trainer with multi-turn agent loops, each prompt may produce a list of `AgentLoopOutput`s. Previously `_compute_score` was called only on the final output, constructing a batch of size 1. This change passes all outputs to `_compute_score` so reward managers can be extended to score intermediate turns, while the default behaviour (scoring only the last output via `data[-1:]`) is preserved for all existing reward managers.

Also fixes a pre-existing bug in `AgentLoopWorkerTQ._agent_loop_postprocess` where `_compute_teacher_logprobs` was called with the raw `output` parameter (which may be a `list[AgentLoopOutput]`) instead of the resolved `final_output`.

### Checklist Before Starting

- [x] Search for similar PRs. Paste at least one query link here: https://github.com/verl-project/verl/pulls?q=is%3Apr+is%3Aopen+multi-output+async+reward+scoring
- [x] Format the PR title as `[{modules}] {type}: {description}`

### Test

Manually ran trainings with:
- `main_ppo` with naive reward manager
- `main_ppo_sync` with single-output rollouts and naive reward manager
- `main_ppo_sync` with multi-output rollouts and a custom reward manager (that combines the outputs into a single sequence in a non-trivial way, and then delegates to the naive reward manager)

All runs could train for a few steps and compute validation without problems.

This behaviour is also covered by current end-to-end tests.

### API and Usage Example

```python
# Before
await self._compute_score(output, prompts=..., responses=..., attention_mask=..., input_ids=..., position_ids=..., kwargs=kwargs)

# After — outputs is list[AgentLoopOutput]
await self._compute_score(outputs, kwargs=kwargs)
```

### Design & Code Changes

- `_compute_score` now accepts `list[AgentLoopOutput]` and internally derives tensors from `outputs[-1]`
- All built-in reward managers (`naive`, `dapo`, `gdpo`, `limited`, `remote`) use `data[-1:]` instead of asserting `len(data) == 1`
- `RewardLoopWorker.compute_score` removes the single-item assertion
- `AgentLoopWorkerTQ._agent_loop_postprocess` simplified and the `_compute_teacher_logprobs` bug fixed
- Updated `docs/advance/reward_loop.rst` to reflect the new interface

### Checklist Before Submitting

- [x] Read the [Contribute Guide](https://github.com/verl-project/verl/blob/main/CONTRIBUTING.md).
- [x] Apply [pre-commit checks](https://github.com/verl-project/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting).
- [x] Add / Update [the documentation](https://github.com/verl-project/verl/tree/main/docs).
- [x] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/verl-project/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: covered by existing end-to-end tests.
- [ ] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ).
- [x] If your PR is related to the `recipe` submodule, please also update the reference to the submodule commit via `git submodule update --remote`: not applicable.

### CODE REVIEW

File: verl/experimental/agent_loop/agent_loop.py [hunk 1]
```diff
@@ -726,27 +718,34 @@ def _compute_position_ids(self, input_ids, attention_mask, multi_modal_inputs) -
...
+        final_output = outputs[-1]
+        if final_output.reward_score is None and enable_async_reward:
             timing = {}
             with simple_timer("compute_score", timing):
+                prompts = torch.tensor(final_output.prompt_ids, dtype=torch.int64).unsqueeze(0)
+                responses = torch.tensor(final_output.response_ids, dtype=torch.int64).unsqueeze(0)
+                input_ids = torch.cat([prompts, responses], dim=1)
+                attention_mask = torch.ones_like(input_ids)
+                multi_modal_inputs = self._compute_multi_modal_inputs(final_output, input_ids.squeeze(0))
+                position_ids = self._compute_position_ids(input_ids, attention_mask, multi_modal_inputs)
```
[the remark this answers is not in the document]
Comment (guillemgt) [reply verl/experimental/agent_loop/agent_loop.py]: Should be fixed now. final_output is used only in order to set the computed reward as the final reward, which the rest of the (pre-existing) code will be treated as a reward for the whole rollout.

File: verl/trainer/main_ppo_sync.py [hunk 2]
```diff
@@ -360,32 +360,13 @@ async def _agent_loop_postprocess(
...
         await self._compute_teacher_logprobs(
-            output,
-            prompt_ids=output.prompt_ids,
-            response_ids=output.response_ids,
+            final_output,
+            prompt_ids=final_output.prompt_ids,
+            response_ids=final_output.response_ids,
             validate=validate,
             sample_kwargs=kwargs,
         )
```
[the remark this answers is not in the document]
Comment (guillemgt) [reply verl/trainer/main_ppo_sync.py:373]: There is a pre-existing comment above indicating that supporting multiple outputs is on the to-do list. This is therefore outside the scope of this PR

**Comment by wuxibin89** [comment]
@guillemgt  Could you also help to update `docs/advance/reward_loop.rst`?

**Comment by guillemgt** [comment]
@wuxibin89 I've updated the documentation

File: verl/experimental/reward_loop/reward_loop.py [hunk 3]
```diff
@@ -144,15 +144,14 @@ async def compute_score_batch(self, data: DataProto) -> list[dict]:
         return outputs
 
     async def compute_score(self, data: DataProto) -> dict:
-        assert len(data) == 1, "RewardLoopWorker only support single data item"
         if self.config.reward.custom_reward_function.path is not None:
             # directly use user-customized reward function
             return await self.reward_manager.run_single(data)
```
Comment (wuxibin89) [inline verl/experimental/reward_loop/reward_loop.py:149]: @yyDing1 will this change break user custom_reward_function? since they may expect only one sample.

Comment (guillemgt) [reply verl/experimental/reward_loop/reward_loop.py:149]: With non-custom reward managers it should be fine, since [they will consider only the last output](https://github.com/verl-project/verl/pull/6228/changes#diff-9d6b9cb7351da6cf2913b8dd90a044c9f8826ba69844cbc359236091bc7d50a9R76) (same behaviour as in current `main`)

Comment (yyDing1) [reply verl/experimental/reward_loop/reward_loop.py:149]: Would it be better to call `compute_score_batch` rather than `compute_score` in the agent loop, so that the code in `reward manager` is not modified?

File: verl/experimental/agent_loop/agent_loop.py [hunk 4]
```diff
@@ -755,9 +771,9 @@ async def _compute_score(self, output, prompts, responses, attention_mask, input
                 )
                 selected_reward_loop_worker_handle = random.choice(self.reward_loop_worker_handles)
                 result = await selected_reward_loop_worker_handle.compute_score.remote(data)
```
Comment (yyDing1) [inline verl/experimental/agent_loop/agent_loop.py:773]: I mean `result = await selected_reward_loop_worker_handle.compute_score_batch.remote(data)` here.

File: verl/experimental/reward_loop/reward_loop.py (hunk 3 above)
Comment (guillemgt) [reply verl/experimental/reward_loop/reward_loop.py:149]: Calling `compute_score_batch` would mean that we are scoring each output _independently_ of others. Each output within the same rollout could have a different reward, which would complicate the advantage estimation in methods like GRPO.
I chose to modify reward managers because that would allow reward functions that need to consider all outputs for a rollout at once (without assuming independence).

File: verl/experimental/agent_loop/agent_loop.py (hunk 4 above)
Comment (yyDing1) [reply verl/experimental/agent_loop/agent_loop.py:773]: or `result = await selected_reward_loop_worker_handle.compute_score.remote(data[-1:])`?

File: verl/experimental/reward_loop/reward_loop.py (hunk 3 above)
Comment (guillemgt) [reply verl/experimental/reward_loop/reward_loop.py:149]: For example, the method in [this paper (Algorithm 1 on page 4)](https://arxiv.org/pdf/2505.20622) (minus the per-batch standardisation of rewards) uses a reward function that needs all outputs in a rollout.

File: verl/experimental/agent_loop/agent_loop.py (hunk 4 above)
Comment (guillemgt) [reply verl/experimental/agent_loop/agent_loop.py:773]: One could do that (in the first case summing the different rewards) but there would be reward functions that aren't possible with this. I have included more details in the other thread, with an example that's possible in the current implementation in the PR by using a custom reward manager.

**Review by wuxibin89 — DISMISSED** [review]

**Comment by wuxibin89** [comment]
@guillemgt cpu_unit_tests ci failed

**Comment by guillemgt** [comment]
@wuxibin89 should be fixed now

**Review by yyDing1 — APPROVED** [review]
